Fix add voucher view to use multi use coupon for bundles.

waheedahmed opened the pull request:
LEARNER-5436

waheedahmed commented:
@edx/learner-spartans @staubina Please review.

staubina commented:
@robrap Please review this ticket.

robrap commented on ecommerce/extensions/basket/tests/test_views.py:
```diff
@@ -787,6 +787,22 @@ def test_voucher_not_valid_for_bundle(self):
+            attribute_type=BasketAttributeType.objects.get(name=BUNDLE),
+            value_text='test_bundle'
+        )
+        self.assert_form_valid_message("Coupon code '{code}' added to basket.".format(code=voucher.code))
```
Do we have separate tests to ensure that the coupon is handled properly (i.e., will be applied correctly to all courses and provide the correct final price)?

robrap commented:
@waheedahmed @staubina: I just saw an email from marketing about 20% off coupon with fine print that it isn't applicable to bundles.  
1. Did they determine this isn't a blocker?
2. Did we ensure (through automated or manual test) that a 20% coupon against a bundle with this fix will apply the 20%, but not apply the bundle offer?

As noted above, although this is approved in order to unblock whatever was blocked, it would be great to get the minor code refactor in (assuming I didn't miscode anything).  It would also be good to determine the proper unit testing around multi-use coupons and bundles, so we don't need a manual test to determine that the bundle offer won't be applied when the multi-use coupon is applied.

waheedahmed commented:
@robrap I did the requested refactoring. I will try to test that locally if I'm unable to test @staubina can you please help me testing this?

robrap commented on ecommerce/extensions/basket/views.py:
```diff
@@ -456,7 +456,10 @@ def apply_voucher_to_basket(self, voucher):
+        is_bundle_purchase = len(bundle_attribute) > 0
+        voucher_program_uuid = voucher.offers.first().condition.program_uuid
+        is_voucher_valid_for_bundle = voucher_program_uuid or voucher.usage == Voucher.MULTI_USE
+        if is_bundle_purchase and not is_voucher_valid_for_bundle:
```
Thanks.  This is much easier to understand.

waheedahmed commented:
@robrap I have added another test case and tried to cover your test scenario in it, please take a look and let me know if this looks good or not?

robrap commented on ecommerce/programs/tests/test_offer.py:
```diff
@@ -68,3 +73,28 @@ def test_offer(self):
         self.assertEqual(basket.total_discount, Decimal(50) * len(lines))
         for line in lines:
             self.assertEqual(line.line_price_incl_tax_incl_discounts, 50)
+
```
I appreciate you writing this, but I’m thinking the assertions right before this code might be enough. I know it is testing Single Use (vs Multi Use), but I think it may be more of a voucher vs offer precedence question. If you are ok with dropping this test, I would do that. If not, let’s discuss.

Chris and I added a minor refactor that came from us trying to understand/clarify some other code.

waheedahmed replied:
@robrap First, I thought same but then I thought we changed the view for multi use vouchers, that's why I added this test. I'm okay with dropping and adding the assertions above L73.

Thanks for the refactor.

waheedahmed commented:
@robrap I have removed the duplicate test and added the assertions above, also did some refactoring with the original test to add voucher using the API call instead of directly adding it into basket.
